packages/language-server/src/plugins/typescript/service.ts

```diff
@@ -350,4 +375,68 @@ async function createLanguageService(
     function getDefaultExclude() {
         return ['node_modules', ...ignoredBuildDirectories];
     }
+
+    /**
+     * Disable usage of project files.
+     * running language service in a reduced mode for
+     * large projects with improperly excluded tsconfig.
+     */
+    function reduceLanguageServiceCapability() {
+        languageService.cleanupSemanticCache();
+        languageServiceReducedMode = true;
+        docContext.notifyExceedSizeLimit?.();
+    }
+}
+
+/**
+ * adopted from https://github.com/microsoft/TypeScript/blob/3c8e45b304b8572094c5d7fbb9cd768dbf6417c0/src/server/editorServices.ts#L1955
+ */
+function getFilenameForExceededTotalSizeLimitForNonTsFiles(
+    compilerOptions: ts.CompilerOptions,
+    tsconfigPath: string,
+    snapshotManager: SnapshotManager
+) {
+    if (compilerOptions.disableSizeLimit) {
+        return;
+    }
+
+    let availableSpace = maxProgramSizeForNonTsFiles;
+    serviceSizeMap.set(tsconfigPath, 0);
+
+    serviceSizeMap.forEach((size) => {
+        availableSpace -= size;
+    });
+
+    let totalNonTsFileSize = 0;
+
+    const filesNames = snapshotManager.getProjectFileNames();
+    for (const fileName of filesNames) {
+        if (hasTsExtensions(fileName)) {
+            continue;
+        }
+
+        totalNonTsFileSize += ts.sys.getFileSize?.(fileName) ?? 0;
```

Comment:
Out of curiosity: Is this performance heavy to check that? How often is this called? My fear is that if this results in many IO operations, that it slows down the intellisense for people.